Honor-riffic: Adds toggleable honorific titles to certain job ID trims

Rhials opened the pull request:
## About The Pull Request

This implements a new, minor, but flavorful system to IDs -- Honorifics. Does going by your full name not suit you? Do you demand respect for the position on Space Station 13 that you've earned and want to be addressed by your title? Do you take yourself WAY too seriously? This is for you.

Toggled by ctrl-clicking your ID, honorifics append a title to your name depending on your position. Certain titles (Captain, Officer, Doctor, etc.) will only append to the start (replacing or including the first/last name), while others (PhD., Esq.) can only be appended at the end.

![image](https://github.com/user-attachments/assets/3ae22744-5f6e-4d12-9ea8-ecd60456b5a9)

Each job TRIM has a set honorific and positions it can be assigned to. A doctor can be "Doctor Peterson" or "Doctor Peterson Bungle" or "Doctor Bungle". A Lawyer can only choose to be "Peterson Bungle, Esq.".

This will only occur when the speaker's voice is the same identity as the one written on the ID's registered name. This should not interfere with Unknown voice obfuscation, stolen ID shenanigans, or anything gameplay-oriented. Hopefully.

This feature is also mononym friendly!

![image](https://github.com/user-attachments/assets/21555023-5dd0-49e0-acd5-2dd0a06ae621)

This also makes `first_name()` and `last_name()` global procs, and adds one to check if a passed string has spaces/dashes/whitespace/whatever.

All of this is compatible with ID name changes, but the voice name must align with the card name to display the honorific. If you are "Peter Stinkypants" with your honorific set to display "Doctor Stinkypants", and your ID's registered name is changed to "Peter Stinker", you show up as "Peter Stinkypants (as Peter Stinker)" with no honorific provided. If you become "Peter Stinker" and have a "Peter Stinker" ID, you will show up as "Doctor Stinker" once again.

That all make sense? Great.

I'm posting this as a draft because I want to make this a roundstart preference, but that is probably going to be a lot of work. This entire PR has been way more work than it was ever meant to be. Argh.

<details>
<summary>So about the ID name stuff...</summary>
<br>
So, when you activate an honorific on your ID, it DOES change the actual object's name. Not the registered name, but the ID's name will go from "Peter Dawson's ID card" to "Captain Dawson's ID card" when an honorific is applied. This, as far as I've tested, does not mess with anything important, but I can totally see it doing so in a way that makes ctrl-Fing through logs harder than it needs to be. This could probably be changed without too much effort so if the issue does arise I can fix it. If not I am totally fine with reverting this PR until I can make it work (if I can at all).
<br>


Admittedly this doesn't have much testing with holopads/radios, but I'm confident the message composure is handled well enough to only display the right names under the right circumstances. If this fucks up logging or naming in any way tell me ASAP because I have a sinking feeling it will in a way more catastrophic than I could ever predict. This PR has been tested thoroughly but I have my limits.
</details>


## Why It's Good For The Game

Having silly little titles to go by is a nice bit of flavor that some players/characters can design around.

These are optional of course, but I fondly recall other servers with similar job-based name displays that I found endearing enough to try out here.

I don't foresee this making chat too noisy. Unless some maniac is constantly swapping title positions, I expect players will decide on an honorific position (if any) and use that for ensuing rounds.
## Changelog
:cl: Rhials
add: You can now ctrl-click IDs to set your Honorific, which will change your display name to reflect your TITLE as a WORKING MAN on SPACE STATION 13.
/:cl:

Rhials committed:
test 1

Rhials committed:
basic stuff works

only works with bare ID -- no PDA. ctrl click to change honorific status/position

had to make firstname/lastname procs global, they're simple regexes so who cares it'll be useful some day

Rhials committed:
damn this is getting sticky

Rhials committed:
why did i do this

Rhials committed:
fixessssss

Rhials committed:
scrape scrape

Rhials committed:
seriously

Rhials committed:
start clean

it always comes back to signals doesn't it

Rhials committed:
more options, better menuing

still doesnt work when in a PDA or wallet or whatever. Argh.

Rhials committed:
even better menuing

Rhials committed:
it sorta works

Rhials committed:
ok now it actually works

ive got to unfuck this signal dude this sucks so bad

Rhials committed:
ok now it actually really works 100% for real this time I promise you

this still overrides names when you have the wrong ID. fuck. Shit. Argh.

Rhials committed:
it didnt actually work wow

Rhials committed:
more titles

apparently those positions I just shittalked are actually important because of the PhD/Esq titles. Huh!

Rhials committed:
other titles

we dont have full-first name honorifics for centcom positions bc sometimes it replaces the name manually

Rhials committed:
"hopefully" nah bro it worked

apparently these also work with the id access manager perfectly fine so cool! im so smart

Rhials committed:
more options

Rhials committed:
mononym safe!

i hate regex

still need to add context tips nvm just did lol

Loafin34 commented:
big fan. now we can have doctor poop be on the station

DrAmazing343 commented:
Captain named Finger

necromanceranne commented:
How does this interact with agent cards and forging?

DrDiasyl commented on code/datums/id_trim/jobs.dm:
```diff
@@ -473,10 +492,15 @@
 		ACCESS_HOP,
 		)
 	job = /datum/job/cook
+	honorific = "Head Chef"
```
Why is "Cook" a Head Chef and actual "Chef" just Chef?

Sadboysuss replied:
Chef is already the head of cooks too

Rhials replied:
![image](https://github.com/user-attachments/assets/964b6078-c495-4315-b291-131fa862d0b7)

Nuts.

DrDiasyl commented on code/datums/id_trim/centcom.dm:
```diff
@@ -94,6 +103,9 @@
 /// Trim for Centcom Commanders. All Centcom and Station Access.
 /datum/id_trim/centcom/commander
 	assignment = JOB_CENTCOM_COMMANDER
+	honorific = "Commander"
```
The random name generator for ERT's make them get "Commander" and such as their actual names, so you probably should remove those from their name pool if you want them to have honorifics as well.

CRITAWAKETS replied:
i too fear Commander Commander

necromanceranne replied:
Sounds like the working name for Cobra Commander.

Rhials replied:
> The random name generator for ERT's make them get "Commander" and such as their actual names, so you probably should remove those from their name pool if you want them to have honorifics as well.

The CentCom honorifics have less selectable positions they can use for this exact reason, but I will confess I haven't yet tested this. If it does actually mess with the names, I'll find a solution.

Exester509 commented:
Why is there a honorific for a Station Engineer but not for the Chief Engineer? Do they just use the same?

SmArtKar commented:
It'd be neat if you could also use shortened versions, like "Dr. Dawson Patterson" instead of "Doctor Dawson Patterson" - same can probably be applies to a lot of other titles (``Capt.``, ``Dir.``, ``Chief`` instead of ``Chief Officer``, etc)

Mothblocks reviewed (commented):
Can you make the honorific_positions a bitfield and only turn it into a list once you need to actually display it? There's about 150 /datum/id_trims on this 60 player Manuel server and that's not a lot but it's some easy waste to clean. It also allows for `::` and `initial` to be used on this.

Mothblocks commented:
Also before anyone asks don't put this on prefs menu until you have a design to pass me, there's a lot of easy ways to make something that looks really bad

Fikou commented:
i dont think this should be a thing for jobs like bitrunner or botanist i dont respect them

Rhials commented:
> It'd be neat if you could also use shortened versions, like "Dr. Dawson Patterson" instead of "Doctor Dawson Patterson" - same can probably be applies to a lot of other titles (`Capt.`, `Dir.`, `Chief` instead of `Chief Officer`, etc)

Honestly I like this idea more too. I did it originally (or at least planned to) but was hesitant about it, probably due to some neuroses involving abbreviations in chat messages that made sense at the time.

> Why is there a honorific for a Station Engineer but not for the Chief Engineer? Do they just use the same?

It's incomplete atm. All titles subject to change.

> How does this interact with agent cards and forging?

Good call, will get back to you on that.

Rhials committed:
adjustment set numero uno

agent ID card doesn't work. Could just add the honorific during the forging process though. Argh.

Rhials committed:
the metamorphosis will be agonizing

still gotta change these from defines arghhhh

Rhials commented:
Made some changes to titles. Added alternates/abbreviations you can select. Not undrafting until the positions are converted to work off of bitfields but feel free to look at the names and make pass your judgement on them.

norsvenska commented on code/datums/id_trim/jobs.dm:
```diff
@@ -824,6 +855,8 @@
 		ACCESS_CMO,
 		)
 	job = /datum/job/paramedic
+	honorifics = list("EMT", "Doctor", "Dr.")
```
paramedics aren't doctors

Rhials replied:
Gonna sidestep a bunch of RPR9 discourse and go with "google says paramedics aren't doctors" instead

Dawnseer commented:
No Doctor Honorific for roboticists? Sad!

Rhials committed:
stop being a doctor

note self: figure out workaround for setting honorific_position_none because every id needs to have the option to remove honorifics. lol./

Rhials committed:
suck a fart out of my ass

SmArtKar commented:
Any progress on this?

Rhials commented:
> Any progress on this?

Holiday season got me busy. I'll probably wrap this up some time this week.

Rhials committed:
dude

Rhials committed:
hahahahahahaha

Rhials committed:
bitifies the rest of the names

MrMelbert commented on code/datums/id_trim/centcom.dm:
```diff
@@ -23,6 +23,8 @@
 	subdepartment_color = COLOR_SERVICE_LIME
 	big_pointer = FALSE
+	honorifics = list("Custodian")
+	honorific_positions = HONORIFIC_POSITION_LAST | HONORIFIC_POSITION_NONE
```
I noticed every single job has `HONORIFIC_POSITION_NONE` set. 

Do you intend on there being jobs that aren't allowed to disable their honorific? 
Or rather do you think there are usecases where a job would be forced to use their honorific? 

If no, I feel like we should just axe the flag and manually include "Disable Honorific" in the input list in ID card code. 
It'd make it a lot easier to write if people could just slap in the flags they want.

MrMelbert replied:
Alternatively you could include "Disable Honorific" by default and instead have a flag `HONORIFIC_POSITION_FORCED` which disables disable honorific

Rhials replied:
> I noticed every single job with a position set also has `HONORIFIC_POSITION_NONE` set.
> 
> Do you intend on there being jobs that aren't allowed to disable their honorific? Or rather do you think there are usecases where a job would be forced to use their honorific?
> 
> If no, I feel like we should just axe the flag and manually include "Disable Honorific" in the input list in ID card code. It'd make it a lot easier to write if people could just slap in the flags they want.

Honestly, I had planned on making it to be a default option so it wouldn't always need to be concluded. After a while though, the potential of a forced honorific felt like a fun idea that could be capitalized on by someone more creative than I. I think I left a comment in the code somewhere mentioning that.

MrMelbert replied:
Then how do you feel about my follow up idea, where you make honorifics have "disableable" in built and you have a separate flag for forced? 
```dm
	honorifics = list("Custodian")
	honorific_positions = HONORIFIC_POSITION_LAST // no none flag, it is implied - you can disable it if you want
```
```dm
	honorifics = list("Judge")
	honorific_positions = HONORIFIC_POSITION_LAST | HONORIFIC_FORCED // force flag meaning you cannot turn it off
```
```
var/list/honorific_options = list()
if(!(honorific_positions & HONORIFIC_FORCED))
    honorific_options += "Disable"
```
IMO this would make it easier to work with in the future but I will leave it up2u

Rhials replied:
I'm fine with just leaving it as-is for now. Double the technical debt and give it to the next person.

MrMelbert commented on code/datums/id_trim/_id_trim.dm:
```diff
@@ -28,6 +28,12 @@
+	var/list/honorifics
+	/// What positions can our honorific take? To prevent names like "Peter Dr."
+	var/honorific_positions = NONE
+	/// What is our selected honorific?
+	var/chosen_honorific
```
Aren't these singletons? I don't think these are supposed to hold state right? So the ID card should store this

MrMelbert commented on code/game/say.dm:
```diff
@@ -147,7 +147,18 @@ GLOBAL_LIST_INIT(freqtospan, list(
+		if(id_slot)
+			var/obj/item/card/id/id_card = id_slot.GetID()
+			if(id_card)
+				SEND_SIGNAL(id_card, COMSIG_MOVABLE_MESSAGE_GET_HONORIFIC, stored_name, carbon_human)
```
This signal seems a bit misleading, as it implies it's a signal sent to movables but it'd only ever sent to id cards

Maybe it should just be a proc `id_card/proc/get_honorifics()`

Rhials replied:
I've renamed the signal to be more specific. I don't remember why I was set on using a signal here (I put up this PR like 3 months ago wtf am I doing), but I really don't want to change anything there right now.

MrMelbert replied:
I think the signal is pretty meh but it can be refactored later, not the worst

MrMelbert commented on code/game/say.dm:
```diff
@@ -147,7 +147,18 @@ GLOBAL_LIST_INIT(freqtospan, list(
+	if(iscarbon(speaker)) //First, try to pull the modified title from a carbon's ID. This will override both visual and audible names.
+		var/mob/living/carbon/carbon_human = speaker
+		var/obj/item/id_slot = carbon_human.get_item_by_slot(ITEM_SLOT_ID)
+		if(id_slot)
+			var/obj/item/card/id/id_card = id_slot.GetID()
+			if(id_card)
+				SEND_SIGNAL(id_card, COMSIG_MOVABLE_MESSAGE_GET_HONORIFIC, stored_name, carbon_human)
```
```suggestion
		var/obj/item/card/id/id_card = id_slot?.GetID()
		if(id_card)
			SEND_SIGNAL(id_card, COMSIG_MOVABLE_MESSAGE_GET_HONORIFIC, stored_name, carbon_human)
```

MrMelbert commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -172,7 +191,7 @@
 
 /obj/item/card/id/dropped(mob/user)
-	UnregisterSignal(user, COMSIG_MOVABLE_POINTED)
+	UnregisterSignal(user, list(COMSIG_MOVABLE_POINTED, COMSIG_MOVABLE_MESSAGE_GET_HONORIFIC))
```
From what I can tell this is never registered on user

MrMelbert commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -871,6 +920,52 @@
 		return insert_money(interacting_with, user) ? ITEM_INTERACT_SUCCESS : ITEM_INTERACT_BLOCKING
 	return NONE
 
+/obj/item/card/id/item_ctrl_click(mob/user)
```
Should only allow this if the id card is being held (so you can still drag them if you feel so inclined)

MrMelbert commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -871,6 +920,52 @@
+		return
+
+	if(trim.honorific_positions == NONE)
+		balloon_alert(user, "no honorific positions! Error!")
```
Probably no need to have a balloon alert if it's a stack trace?

MrMelbert commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -871,6 +920,52 @@
+		honorific_title = null
+		balloon_alert(user, "honorific disabled")
+	else
+		trim.chosen_honorific = tgui_input_list(user, "what honorific do you want to use?", "Flair!!!", trim.honorifics)
```
```suggestion
		trim.chosen_honorific = tgui_input_list(user, "What honorific do you want to use?", "Flair!!!", trim.honorifics)
```

MrMelbert replied:
Same here

MrMelbert commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -871,6 +920,52 @@
+		if(trim.honorific_positions & readable_names[i]) //If the positions list has the same bit value as the readable list.
+			choices += i
+
+	var/chosen_position = tgui_input_list(user, "what position do you want your honorific in?", "Flair!", choices)
```
```suggestion
	var/chosen_position = tgui_input_list(user, "What position do you want your honorific in?", "Flair!", choices)
```

MrMelbert replied:
You need to check after the input the user is still alive and in range of the id card and etc etc

Rhials committed:
by the end of this, this shouldnt happen at all

Rhials committed:
capitalism

Rhials committed:
home stretch

Rhials committed:
finish line

MrMelbert commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -871,6 +923,54 @@
+			choices += i
+
+	var/chosen_position = tgui_input_list(user, "What position do you want your honorific in?", "Flair!", choices)
+	if(user.stat < CONSCIOUS || !in_contents_of(user))
```
We generally use `user.incapacitated` rather than `user.stat < ...`

Rhials committed:
incapacitated check

MrMelbert commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -858,6 +890,26 @@
 
+/// Re-generates the honorific title. Returns the compiled honorific_title value
+/obj/item/card/id/proc/update_honorific()
+	var/is_mononym = FALSE
+	if(is_mononym(registered_name))
+		is_mononym = TRUE
```
```suggestion
	var/is_mononym = is_mononym(registered_name)
```

MrMelbert commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -172,7 +194,7 @@
 
 /obj/item/card/id/dropped(mob/user)
-	UnregisterSignal(user, COMSIG_MOVABLE_POINTED)
+	UnregisterSignal(user, list(COMSIG_MOVABLE_POINTED))
```
```suggestion
	UnregisterSignal(user, COMSIG_MOVABLE_POINTED)
```

MrMelbert commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -871,6 +923,54 @@
+	if(honorific_position_to_use & HONORIFIC_POSITION_NONE)
+		balloon_alert(user, "honorific disabled")
+	else
+		chosen_honorific = tgui_input_list(user, "What honorific do you want to use?", "Flair!!!", trim.honorifics)
+		if(!chosen_honorific || user.incapacitated || !in_contents_of(user))
+			return
+		switch(honorific_position_to_use)
```
This has to be done otherwise the sanity check is pointless (as it updates the var regardless of if the mob can) 
```suggestion
		var/new_honorific = tgui_input_list(user, "What honorific do you want to use?", "Flair!!!", trim.honorifics)
		if(!new_honorific || user.incapacitated || !in_contents_of(user))
			return
		chosen_honorific = new_honorific
		switch(honorific_position_to_use)
```

MrMelbert committed:
Assuming direct control
